### DISCUSSION TITLE
Specify a different build directory for Rust projects

### DISCUSSION THREAD

**Discussion by twiclo** [VonHeikemen/lsp-zero.nvim#134] [Q&A]
I'd like my lsp to use a different build dir so I'm don't get the "Waiting on build directory lock" when trying to run `cargo run` manually from a terminal. Is this possible?

**Comment by VonHeikemen** [comment]
This question is very specific to `rust-analyzer`. I don't have the answer.

**Reply by twiclo** [reply]
There's an env variable that you can set to change the build dir. Do you think setting that variable through `.build_options` would work?

**Reply by VonHeikemen** [reply]
Probably. There is an option called `cmd_env`, is a list of key/value pairs. You can try that.

```lua
lsp.build_options('rust-analyzer', {
  cmd_env = {SOME_VAR = 'some value'}
})
```

**Reply by twiclo** [reply]
```
local rust_lsp = lsp.build_options('rust_analyzer', {
  cmd_env = { 'CARGO_TARGET_DIR' = '~/.build' }
})
require('rust-tools').setup({ server = rust_lsp })
```
Getting this error:
`E5107: Error loading lua [string ":source (no file)"]:3: '}' expected near '='`

**Reply by VonHeikemen** [reply]
If you want to use the quotes then you need to add square brackets. And also, I recommend you "expand" the path, just to be safe.

```lua
cmd_env = { ['CARGO_TARGET_DIR'] = vim.fn.expand('~/.build') }
```

**Reply by twiclo** [reply]
Alright here's what I have now
```
local rust_lsp = lsp.build_options('rust_analyzer', {
  cmd_env = { 'CARGO_TARGET_DIR' = '~/.build' }
})
require('rust-tools').setup({ server = rust_lsp })
```
Says I'm missing module rust-tools

**Reply by VonHeikemen** [reply]
Did you install [rust-tools.nvim](https://github.com/simrat39/rust-tools.nvim)?

You can use `.configure` to pass options to a language server. This should work without rust-tools.

```lua
lsp.configure('rust-analyzer', {
  cmd_env = { ['CARGO_TARGET_DIR'] = vim.fn.expand('~/.build') }
})
```

(this one should be called before `lsp.setup()`)
